server.py

```diff
@@ -61,6 +64,24 @@ def read_file():
     ip_address = request.remote_addr if request.args.get('ip') is None else request.args.get('ip')
     metadata = getattr(g, 'metadata', None)
     filename = request.args.get('uuid')
+    if app.config['use_dist_replication']:
+        concurrent_requests = metadata.add_concurrent_request(file_uuid)
+        if concurrent_requests is not None:
+            # Make sure that the number of concurrent requests is under k.
+            # If not, replicate to another server.
+            if int(concurrent_requests) >= app.config['k']:
+                # 1) Find the closest server.
+                target_server = metadata.get_closest_server()
+                # 2) Copy the file to that server.
+                clone_file(request.args.get('uuid'), target_server, 'DISTRIBUTE_REPLICATE', ip_address)
+        else:
+            raise Exception('Something fishy is going on... Should have at least one request')
+
+        # remove the number of concurrent requests to the file
+        @after_this_request
```

Comment:
i think this is a decorator which is basically a function wrapper. it adds some salt to remove_request() thus mapping it to a more specific function. in this case, seems remove_request() is added as a callback function.